scvi/data/fields/_obs_field.py

```diff
@@ -110,30 +118,34 @@ class CategoricalObsField(BaseObsField):
     CATEGORICAL_MAPPING_KEY = "categorical_mapping"
     ORIGINAL_ATTR_KEY = "original_key"
 
-    def __init__(self, registry_key: str, obs_key: Optional[str]) -> None:
+    def __init__(
+        self, registry_key: str, obs_key: Optional[str], mod_key: Optional[str] = None
+    ) -> None:
         self.is_default = obs_key is None
         self._original_attr_key = obs_key or registry_key
-        super().__init__(registry_key, f"_scvi_{registry_key}")
+        super().__init__(registry_key, f"_scvi_{registry_key}", mod_key=mod_key)
 
         self.count_stat_key = f"n_{self.registry_key}"
 
     def _setup_default_attr(self, adata: AnnData) -> None:
         self._original_attr_key = self.attr_key
-        adata.obs[self.attr_key] = np.zeros(adata.shape[0], dtype=np.int64)
+        mod_adata = self._maybe_get_modality(adata)
+        self.set_field_data(adata, np.zeros(mod_adata.n_obs, dtype=np.int64))
```

Comment:
should this be set in the mod_data?

Reply:
I see now that it is, maybe ambiguous from the way this is coded....

Reply:
yea this is what i dont like about this approach. `adata` can either be mudata or anndata at any point which can be very confusing